Handle matrices in Add and Mul as a postprocessor

asmeurer opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

Previously matrices were absorbed into the coeff, but this led to several
bugs, including some that had wrong results.

I still need to see how this impacts https://github.com/sympy/sympy/pull/15121. Ideally both this and that should get merged before the next release. 

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234". See
https://github.com/blog/1506-closing-issues-via-pull-requests . Please also
write a comment on that issue linking back to this pull request once it is
open. -->


#### Brief description of what is fixed or changed

Fixes #15528.
Closes #15666 
Fixes #15665
Closes https://github.com/asmeurer/sympy/pull/76
Closes #15692 

#### Other comments

`Add(scalar, matrix)` remains unevaluated (it already works this way in master). Without that, something like `(x + y).subs({x: matrix, y: matrix})` cannot work without the `simultaneous=True` flag, as it creates an intermediate expression `Add(matrix, y)`. Other algorithms break as well, such as `factor`.

#### Release Notes

<!-- Write the release notes for this release below. See
https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more information
on how to write release notes. The bot will check your release notes
automatically to see if they are formatted correctly. -->

*(Note: I'm not sure if this actually changes any behavior since 1.3. #15528 and #15665 were both regressions in master.)*

<!-- BEGIN RELEASE NOTES -->
* core
  * matrices are now handled in Add/Mul using postprocessors. 
  * the experimental `_constructor_postprocessor_mapping` API no longer applies postprocessors from subexpressions. Note that this API is experimental and may change or be removed in a future version of SymPy without deprecation. 
<!-- END RELEASE NOTES -->

asmeurer committed:
Handle matrices in Add and Mul as a post processor

Previously matrices were absorbed into the coeff, but this led to several
bugs, including some that had wrong results.

There is a slight regression here, as seen from the XFAIL test in test_subs.
Namely, substituting multiple matrices into a scalar Add expression will raise a
TypeError because it tries computes the subexpressions. For instance, (x +
y).subs({x: Matrix(...), y: Matrix(...)}). This can be worked around by using
the simultaneous=True flag to subs.

Fixes #15528.

asmeurer committed:
Keep Add(scalar, matrix) unevaluated

scalar + matrix still raises a TypeError.

This allows algorithms to substitute matrices with noncommutative dummy
symbols and manipulate them. subs in particular substitutes each term
iteratively (unless simultaneous=True), so something like (x + y).subs({x:
matrix, y: matrix}) would raise a TypeError without this change. See also the
previous commit message.

asmeurer committed:
Correctly handle the case of no matrices in the MatrixExpr postprocessors

For some reason items get cached in the post-processor and it causes them to
be called on Adds and Muls that don't have any top-level MatrixExprs.

asmeurer commented:
@Upabjojr why do instances get added to the Basic._constructor_postprocessor_mapping? It causes some unpredictable behavior, where the postprocessors only get called if you create certain other expressions first. It also seems to me that it is a major memory leak, since there is nothing that ever removes them. 

I also noticed that nothing in the code at this point in time uses the postprocessors? I thought there was something that used them. Were they removed at some point?

asmeurer commented:
I guess the idea is for expressions to be able to manipulate things higher up in the tree than one level. For example, one of the examples in the tests, `w = SymbolRemovesOtherSymbols('w')` removes other symbols in a Mul, and should work for both `x*w` and `x*w**2`. 

I think this is overkill. If we really want to allow that, we should check for subexpressions, rather than saving every expression that is seen in a dictionary forever. But even that is expensive, and I don't know if there is a valid use-case for it (again, outside of this PR currently there are no actual postprocessors used anywhere other than the tests). 

So given that this will create a nasty memory leak with this PR, and likely some undefined behavior as well, and given that the API is experimental, I think we should remove it. If we want something like this we need to think harder about a better way to do it. 

It certainly isn't needed for this PR---we only need Adds and Muls with matrices in their args to become MatAdd and MatMul. It doesn't matter if matrices are in subexpressions (like a MatrixElement), only if the top-level args have matrices. Honestly, postprocessors are overkill for this. We should probably refactor the flatten code to generalize the postprocessor I've written here, so that an Add or Mul can return a different class.

asmeurer committed:
Remove the ability for constructor postprocessors to work with subexpressions

This worked by storing objects in the
Basic._constructor_postprocessor_mapping. However, they were never removed
from that mapping, meaning it could be a memory leak. Furthermore, the fact
that they were there forever meant that some expressions would get
postprocessors called only after other expressions were created, resulting in
nondeterministic undefined behavior.

asmeurer commented:
I've gone ahead and removed it for now. This should also improve the performance of the postprocessors. 

I should probably revisit https://github.com/sympy/sympy/pull/13161 now that I better understand how the postprocessors work. I wonder if this subexpression issue was part of my major concerns that I had there.

asmeurer committed:
Remove confusing comment

Now that the postprocesors don't handle subexpressions, the comment isn't
valid. However, I think it is worth leaving the logic there, just in case it
somehow gets triggered still.

jksuom commented:
I have tried to understand why matrices end up among the arguments of Add and Mul in the first place. There seem to be explicit tests with matrix arguments, but is that all there is? Would it be possible to have only core objects as arguments? (Plus possibly AccumBounds, and even those could perhaps be moved to core.)

asmeurer commented:
I think the most common way is through subs, like `(x + y).subs({x: matrix, y: matrix})`. It seems to boil down to that when using simplification functions like `factor` on matrix expressions, from what I remember. 

For Mul at least, it could also just happen because someone uses `Mul` instead of `*` for an expression containing matrices. For instance, someone mentioned [on Gitter](https://gitter.im/sympy/sympy?at=5c5a61d5ca428b064560e7ea) that they are doing this. That also seemed to be the case for the issue https://github.com/sympy/sympy/issues/15528.

jksuom commented:
Substitutions can produce expression that appear to make no sense.
```
>>> p = Poly(x)
>>> q = Quaternion(1, 1, 1, 1)
>>> (x + y).subs(y, q).subs(x, p)
Poly(x, x, domain='ZZ') + 1 + 1*i + 1*j + 1*k
>>> type(_)
<class 'sympy.core.add.Add'>
```
It seems to me that matrices are not the only non-core expressions that should be taken care of. Perhaps all such arguments of Add should be appended to `extra`, not only matrices.

jksuom commented:
How about changing the signature of `_from_args` to `_from_args(cls, args, ncargs)`. The object would be created from `args`, and postprocessing would be done by combining the result with the sequence of  `ncargs` one at a time. The private methods of each ncarg would take care of the necessary type checks and coercions.

asmeurer commented:
> Substitutions can produce expression that appear to make no sense.

Yes, but it can be necessary for intermediate expressions. See my commit message on 7025957.

> How about changing the signature of `_from_args` to `_from_args(cls, args, ncargs)`. The object would be created from `args`, and postprocessing would be done by combining the result with the sequence of `ncargs` one at a time. The private methods of each ncarg would take care of the necessary type checks and coercions.

I'll have to think about that.

asmeurer commented:
I guess the postprocessors that were there were on Quantity (units), but were removed in https://github.com/sympy/sympy/pull/14923. It was only used to do dimension checking, raising an exception. But for the reasons discussed there and at https://github.com/sympy/sympy/issues/14547, it was a bad idea to do dimension checking at expression creation, because it makes checks throughout SymPy not work, in this case, a `x - 1` check in `log`. I think the same applies here for Add and MatrixExpr.

asmeurer committed:
Remove explicit check for MatrixExpr in Mul.flatten

They already get added to nc_part anyway.

asmeurer commented:
One thing that is not so nice about this is that we still cannot remove the explicit mentions of MatrixExpr from Mul or Add. That is because we need special behavior for `0` in the case of Mul and nan and infinity in the case of Add. Otherwise things like `Mul(0, matrix)` would return the scalar `0` in every case. 

Perhaps this API would be better suited as a preprocessor rather than a postprocessor.

asmeurer commented:
I think the best solution will require some sort of [multiple dispatch](https://github.com/sympy/sympy/pull/14490#issuecomment-384095425). However, that requires some more thought/work than I would like to do for this PR. This fixes a release blocking issue, so since the API exists now and it works, I think we should go with this as-is.

asmeurer committed:
Make MatMul.args_cnc return a list instead of a tuple

This matches Mul.args_cnc, and is needed for gcd_terms().

asmeurer committed:
Fix gcd_terms with matrix expressions

Mul._from_args can only be used if you already know Mul has processed the
args. If it is an arbitrary expression, or if you didn't check if that
expression is actually a Mul (as opposed to a MatMul), it cannot be used.

asmeurer commented:
Starting to test this with https://github.com/sympy/sympy/pull/15121 and the issues mentioned there. I fixed one issue with gcd_terms. If you merge this branch with that one, factor no longer errors. However:

```py
>>> import sympy as sy
>>> n = sy.symbols('n')
>>> A = sy.MatrixSymbol("A",n,n)
>>> B = sy.MatrixSymbol("B",n,n)
>>> C = sy.MatrixSymbol("C",n,n)
>>> M = A.inverse()*B.inverse() - A.inverse()*C*B.inverse()
>>> a = B.inverse()*M.inverse()*A.inverse()
>>> factor(a)
B**(-1)*(((1 - C)*B**(-1))*A**(-1))**(-1)*A**(-1)
>>> a
B**(-1)*(A**(-1)*B**(-1) - A**(-1)*C*B**(-1))**(-1)*A**(-1)
>>> pprint(a)
                        -1
 -1 ⎛ -1  -1   -1    -1⎞    -1
B  ⋅⎝A  ⋅B   -A  ⋅C⋅B  ⎠  ⋅A
>>> pprint(factor(a))
                     -1
 -1 ⎛         -1  -1⎞    -1
B  ⋅⎝(1 - C)⋅B  ⋅A  ⎠  ⋅A
```

This is wrong. It reordered the terms.

asmeurer committed:
Merge branch 'master' into matrices-postprocessor

asmeurer commented:
Another issue here is that `evaluate=False` will skip the post processors. So something like `gcd_terms(2*A*B + 2*B*A)`, which returns `Mul(2, A*B + B*A, evaluate=False)` will not return a MatMul.

asmeurer committed:
Call the postprocessors even when evaluate=False

asmeurer committed:
Fix gcd_terms to return a MatMul on matrix expressions

This requires using Mul(evaluate=False) instead of Mul._make_args, so that the
postprocessors get called.

asmeurer commented:
I've just found another problem. `factor_nc` works by substituting matrices with noncommutative dummies (`_mask_nc`). I'm not entirely clear why. 

The problem is that MatMul is designed to pull out non-matrices to the front, meaning `MatMul(Matrix, NoncommutativeSymbol, Matrix)` will reorder the terms. Keeping them in order is not easy because it assumes that anything not in the factor is a matrix. 

I don't know what the best way to fix this is.

asmeurer committed:
Fix MatMul.args_cnc() tests

asmeurer committed:
Keep noncommutative non-Matrix arguments in order in MatMul.args_cnc()

asmeurer commented:
@smichr do you know why `_mask_nc` is used instead of operating on the expressions directly? It looks like it was added by you in 7dde2b2a87c.

asmeurer commented:
Oh I didn't even notice that it is using `1 - C` instead of `I - C`. I think `factor_nc` will need to be modified to explicitly handle matrix expressions. 

I think to avoid wrong results, we should just make MatMul raise an exception if created with noncommutative scalars for now.

asmeurer committed:
Set Trace.is_commutative = True

asmeurer committed:
Disallow noncommutative scalars in MatMul

The problem is that some algorithms try to replace matrices with
noncommutative Dummys. But this only works if it can keep them in the right
order, which presently it doesn't, as it separates out the scalar and matrix
parts. Since the current behavior can lead to wrong results, e.g., with
factor_nc(), it is better to explicitly disallow it.

This breaks factor_nc() from working on matrix expressions. However, it only
worked correctly in some cases. For example,
`factor_nc(B**(-1)*(A**(-1)*B**(-1) - A**(-1)*C*B**(-1))**(-1)*A**(-1))`
returned `B**(-1)*(((1 - C)*B**(-1))*A**(-1))**(-1)*A**(-1)`, which not only
reordered the terms incorrectly, but used 1 instead of I.

asmeurer commented:
I've pushed a commit that disallows noncommutatives in MatMul. This makes factor_nc not work at all with matrices anymore. Previously, it worked, but also returned wrong results in some cases (like the above example). I haven't checked the other simplification functions yet. 

`factor_nc` will need to be rewritten to explicitly work with matrices. Not only is substituting noncommutative Dummys something that currently cannot work, but it needs to use IdentityMatrix instead of 1. I don't think shape mismatch should be an issue. Unless I'm missing something, I believe any valid noncommutative ring identity should have the same shape match characteristics on both sides (e.g., `A*(B + C)` and `A*B + A*C` are either both valid or both invalid shape-wise for matrices `A`, `B`, and `C`). 

 I think it should work to substitute matrices for noncommutatives, perform the factor, and then replace them back and replace constant additive factors with identity matrices. But that substitution needs to happen in a better way than it is done now, so that you don't have noncommutative Dummys in MatMul. I don't plan on implementing this myself, but it should be straightforward if someone else wants to give it a go.

asmeurer committed:
Better error message

asmeurer committed:
Add an XFAIL test for a matrix expression that could give a wrong result

That is, it would give a wrong result if noncommutative scalars were allowed
in MatMul.

asmeurer committed:
Add a test for diff(Trace(A)*A, A)

See #15121.

asmeurer commented:
Most of the simplification functions now raise NotImplementedError. The exception is `cancel`, which fails here: https://github.com/sympy/sympy/blob/946be880abad6186c74eb3a10d608bebbcaf0851/sympy/polys/polytools.py#L6622. First off `f.func._from_args(c)` has to be replaced with `f.func(*c)`. 

But then there is another fundamental problem, which is that `MatAdd()` with no arguments cannot return anything meaningful. Right now it remains unevaluated, but then `.shape` errors. If you make it return `S.Zero` then other places fail because that isn't a matrix. 

The best I can think of is to make `MatAdd()` return `S.Zero` and `MatAdd(0, Matrix)` remove the 0 without error, which is not ideal, but it's the only way to make things like the above work without error.

Maybe we need some kind of special matrix shape object that isn't type checked, so you could create an n x m zero matrix that allows `n` and `m` to match any shape they are tested against.

jksuom commented:
Is there some place where objects like `MatAdd()` are naturally wanted? Otherwise, we could probably make them raise an exception.

asmeurer commented:
We want to be able to manipulate matrix expressions like any normal noncommutative expression. In particular simplification functions like `factor` and `expand` should work.

asmeurer commented:
> Maybe we need some kind of special matrix shape object that isn't type checked, so you could create an n x m zero matrix that allows `n` and `m` to match any shape they are tested against.

I gave this idea an attempt at https://github.com/sympy/sympy/pull/15986. Combined with this branch, `cancel` does work, in that it doesn't raise an exception at least, although it returns an expression with a bunch of unevaluated `0` and `I` matrices. I'll need to check why those don't get automatically evaluated out.

asmeurer commented:
Any objections to this? This fixes a critical bug in master (wrong results from matrix expressions), and is blocking a release. @Upabjojr do you have any comments?

asmeurer committed:
Merge branch 'master' into matrices-postprocessor

Upabjojr commented:
Did you try some benchmarking? I fear that postprocessors may slow down the code.

asmeurer commented:
I'll do that. Note that I've removed the "nested expression" part of the postprocessors, so that should speed things up quite a bit from the way it was. 

Hopefully, it shouldn't affect the performance of expressions that don't have matrix expressions, but that should definitely be checked against.

asmeurer commented:
If someone could help me benchmark this that would be great. I have a hard time getting consistent results on my computer. Note that if you use something other than asv and the benchmark suite that Add and Mul are cached, so you have to turn the cache off to get real benchmark values.

Can someone run `asv` on the benchmarks repo on master vs. this branch? Also test my `postprocessor-performance` branch, which is this branch plus a commit that may be an optimization (I'm not sure). Make sure to pull the benchmarks repo so that https://github.com/sympy/sympy_benchmarks/pull/47 is included.

asmeurer committed:
Merge branch 'master' into matrices-postprocessor

asmeurer committed:
Remove an unused import

Upabjojr commented on sympy/matrices/expressions/tests/test_matexpr.py:
```diff
@@ -342,7 +341,7 @@ def test_inv():
     assert B.inv() == B**-1
 
-
+@XFAIL
```
you are adding some `XFAIL` s, are you sure it's a good idea?

asmeurer replied:
There are two XFAILs added here. I have mentioned the reasoning in the commit messages, but I will reiterate here. 

This XFAIL has to do with functions like `factor` no longer working with matrix expressions. In fact, many simplification functions now raise `NotImplementedError("noncommutative scalars in MatMul are not supported.")` (in some instances, you have to merge https://github.com/sympy/sympy/pull/15121 to see this message). This message results from algorithms that try to substitute matrix symbols with noncommutative symbols. This is fundamentally broken, because MatMul separates noncommutative symbols as the coefficient, effectively reordering them. You can see the commit message of https://github.com/sympy/sympy/pull/15948/commits/b98319aa86bdc2917e6fe97906944e2556478a43 an example where `factor` was returning a wrong result. I decided it was better for these functions to raise NotImplementedError rather than give wrong results in some cases. Basically, each algorithm will need to be written explicitly with matrix expressions in mind, or else some higher level fix would be needed (maybe make MatMul.subs(Matrix, noncommutative Symbol) do something smarter). Most of these already don't work in master because of https://github.com/sympy/sympy/pull/15121 (see https://github.com/sympy/sympy/issues/15120 and https://github.com/sympy/sympy/issues/6249), so I don't see an issue with XFAILing these for now. 

The other XFAIL is below. It is that Add(nan, matrix) returns nan. This is because of the way Add does its nan short circuiting. It didn't deem it worthy to make this work, since it would require refactoring Add.flatten (right now it unconditionally returns nan whenever it sees a nan). If it ever becomes an issue, we can go in and fix it.

asmeurer replied:
Oh and I forgot the third XFAIL in the postprocessor tests. This is because I removed the ability for postprocessors to work on subexpressions. This is necessary for performance purposes. The current implementation also produces a memory leak. Since the API is not used anywhere else in library code, and it is explicitly designated as experimental, I think we can make such changes to it. 

Actually I have another suggestion for an even more dramatic change to the API [here](https://github.com/asmeurer/sympy/commit/b0777d925d005bf5973702033a7147bda09a1b64). I haven't pushed it to this branch yet because I want to know if it is actually more performant or not. If it isn't, it likely isn't worth the change (even though the API is itself much cleaner IMO).

asmeurer commented:
Any further comments here? 

I just ran the benchmarks on my computer. I still don't trust the asv results from my computer a whole lot, but the results I got was that this branch is slightly faster than master (actually f8e46e9e74, the most recent `master` commit that has been merged into this branch). It is also faster than `postprocessor-performance`.

smichr commented:
> why `_mask_nc` is used instead of operating on the expressions directly.

I *think* it might be that I didn't have a good method of factoring non-commutatives intelligently other than removing prefixes and suffixes, so I let factor do the work and check permutations at the end to see if something worked.

smichr committed:
typo

smichr commented:
This is a level of wrangling above what I am accustomed to thinking about, so I'm not sure that I can see enough implications to comment meaningfully.

asmeurer commented:
I'm reasonably sure the changes themselves are fine here. My main concern is with the performance, since this affects the core (the postprocessors will be checked on any Add or Mul that is created, though not actually called unless they include matrix expressions).

If someone wants to verify my benchmarks here that would help. Here are the steps to do it

1. Install `asv` (with pip or conda)
2. Clone https://github.com/sympy/sympy_benchmarks. If you already have it cloned make sure you pull the latest master.
3. In sympy_benchmarks repo, edit `asv.conf.json` and change `repo` ([this line](https://github.com/sympy/sympy_benchmarks/blob/94882c2720ff06936a7990d6a7c4f68ced4bcd50/asv.conf.json#L14)) to the local path to your `sympy` clone.
4. Make sure you have my remote added and fetched on your sympy clone.
5. From the sympy_benchmarks repo, run

```
asv run f8e46e9e74 -s 1 # master
asv run c825dd5bd3eea83d22d5d226cf553f9960075eba -s 1 # matrices-postprocessor
asv run b0777d925d005bf5973702033a7147bda09a1b64 -s 1 # postprocessor-performance
```

The first command will require you to answer some questions about your machine. 

Make sure nothing else is running on your computer while you run these benchmarks. This will take about 20-30 min. to finish, depending on how fast your computer is. 

6. Run

```
asv compare f8e46e9e74 c825dd5bd3eea83d22d5d226cf553f9960075eba
asv compare c825dd5bd3eea83d22d5d226cf553f9960075eba b0777d925d005bf5973702033a7147bda09a1b64
```

and report the results here.

asmeurer commented:
@smichr in this branch, substituting noncommutative symbols in a MatMul results in NotImplementedError (https://github.com/sympy/sympy/pull/15948/files#diff-557d967fd32b8494f5b26ffcbc1dc4a4R92). The reason is that MatMul stores the coefficient and the matrix separately, and noncommutative symbols must necessary be part of the coefficient (they don't have a shape). Thus, this reorders the terms, which has resulted in some wrong results (see https://github.com/sympy/sympy/pull/15948/commits/b98319aa86bdc2917e6fe97906944e2556478a43). As a result, almost all simplification functions now produce NotImplementedError on matrix expressions in this branch (many of them already don't work in master because of https://github.com/sympy/sympy/pull/15121).

My guess is that we can fix this wholesale by improving `_mask_nc` or maybe even `subs` itself, but I didn't do anything about that here.

smichr commented:
OK, improvement of `_mask_nc`/`subs` need not delay this. Nobody has confirmed performance in 4 days so this will be committed.
